File: mulle/rules/validation.py
"""Validation-funktioner för spelarhandlingar i Mulle."""

from ..models.board import Board
from ..models.player import Player


class InvalidAction(Exception):
    """Undantag som höjs när en spelaråtgärd inte är tillåten."""
    pass


def player_has_builds(board: Board, player: Player) -> bool:
    """
    Kontrollerar om en spelare har några byggen (låsta eller olåsta) på bordet.
    
    Args:
        board: Spelbrädet
        player: Spelaren att kontrollera
        
    Returns:
        True om spelaren har minst ett bygge, False annars
    """
    for build in board.list_builds():
        if build.owner == player.name:
            return True
    return False


def ensure_can_trail(board: Board, player: Player, card: Card = None) -> None:
    """
    Kontrollerar om en spelare får 'släppa' (trail) ett kort till bordet.
    
    En spelare får INTE släppa kort om spelaren har ett eller flera byggen
    på bordet. Spelaren måste först ta in (capture) sina byggen.
    
    Args:
        board: Spelbrädet
        player: Spelaren som vill släppa kort
        
    Raises:
        InvalidAction: Om spelaren har byggen på bordet
    """
    if player_has_builds(board, player):
        if card:
            raise InvalidAction(
                f"Kan inte släppa {card.code()} - {player.name} har byggen på bordet som måste tas in först!"
            )
        else:
            raise InvalidAction(
                f"Kan inte släppa kort: {player.name} har byggen på bordet som måste tas in först"
            )

File: src/rules/validation.ts
import { Board } from '../models/Board';
import { Player } from '../models/Player';
import { Card } from '../models/Card';

/**
 * Exception thrown when a player action is not allowed
 */
export class InvalidAction extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'InvalidAction';
  }
}

/**
 * Check if a player has any builds on the board
 */
export function playerHasBuilds(board: Board, player: Player): boolean {
  for (const build of board.listBuilds()) {
    if (build.owner === player.name) {
      return true;
    }
  }
  return false;
}

/**
 * Ensure that a player can trail (discard to table).
 * A player cannot trail if they have builds on the board.
 */
export function ensureCanTrail(
  board: Board,
  player: Player,
  card?: Card
): void {
  if (playerHasBuilds(board, player)) {
    if (card) {
      throw new InvalidAction(
        `Kan inte släppa ${card.code()} - ${player.name} har byggen på bordet som måste tas in först!`
      );
    } else {
      throw new InvalidAction(
        `Kan inte släppa kort: ${player.name} har byggen på bordet som måste tas in först`
      );
    }
  }
}
